File: py/range.go
// Copyright 2018 The go-python Authors.  All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

// Range object

package py

// A python Range object
// FIXME one day support BigInts too!
type Range struct {
	Start  Int
	Stop   Int
	Step   Int
	Length Int
}

// A python Range iterator
type RangeIterator struct {
	Range
	Index Int
}

var RangeType = NewTypeX("range", `range(stop) -> range object
range(start, stop[, step]) -> range object

Return a virtual sequence of numbers from start to stop by step.`,
	RangeNew, nil)

var RangeIteratorType = NewType("range_iterator", `range_iterator object`)

// Type of this object
func (o *Range) Type() *Type {
	return RangeType
}

// Type of this object
func (o *RangeIterator) Type() *Type {
	return RangeIteratorType
}

// RangeNew
func RangeNew(metatype *Type, args Tuple, kwargs StringDict) (Object, error) {
	var start Object
	var stop Object
	var step Object = Int(1)
	err := UnpackTuple(args, kwargs, "range", 1, 3, &start, &stop, &step)
	if err != nil {
		return nil, err
	}
	startIndex, err := Index(start)
	if err != nil {
		return nil, err
	}
	if len(args) == 1 {
		length := computeRangeLength(0, startIndex, 1)
		return &Range{
			Start:  Int(0),
			Stop:   startIndex,
			Step:   Int(1),
			Length: length,
		}, nil
	}
	stopIndex, err := Index(stop)
	if err != nil {
		return nil, err
	}
	stepIndex, err := Index(step)
	if err != nil {
		return nil, err
	}
	length := computeRangeLength(startIndex, stopIndex, stepIndex)
	return &Range{
		Start:  startIndex,
		Stop:   stopIndex,
		Step:   stepIndex,
		Length: length,
	}, nil
}

func (r *Range) M__getitem__(key Object) (Object, error) {
	if slice, ok := key.(*Slice); ok {
		return computeRangeSlice(r, slice)
	}

	index, err := Index(key)
	if err != nil {
		return nil, err
	}
	index = computeNegativeIndex(index, r.Length)

	if index < 0 || index >= r.Length {
		return nil, ExceptionNewf(IndexError, "range object index out of range")
	}
	result := computeItem(r, index)
	return result, nil
}

// Make a range iterator from a range
func (r *Range) M__iter__() (Object, error) {
	return &RangeIterator{
		Range: *r,
		Index: r.Start,
	}, nil
}

func (r *Range) M__str__() (Object, error) {
	return r.M__repr__()
}

func (r *Range) M__repr__() (Object, error) {
	return r.repr()
}

func (r *Range) M__len__() (Object, error) {
	return r.Length, nil
}

// Range iterator
func (it *RangeIterator) M__iter__() (Object, error) {
	return it, nil
}

// Range iterator next
func (it *RangeIterator) M__next__() (Object, error) {
	r := it.Index
	if it.Step >= 0 && r >= it.Stop {
		return nil, StopIteration
	}

	if it.Step < 0 && r <= it.Stop {
		return nil, StopIteration
	}
	it.Index += it.Step
	return r, nil
}

func computeItem(r *Range, item Int) Int {
	incr := item * r.Step
	res := r.Start + incr
	return res
}

func computeRangeLength(start, stop, step Int) Int {
	var lo, hi Int
	if step > 0 {
		lo = start
		hi = stop
		step = step
	} else {
		lo = stop
		hi = start
		step = (-step)
	}

	if lo >= hi {
		return Int(0)
	}
	res := (hi-lo-1)/step + 1
	return res
}

func getIndexWithDefault(i Object, d Int) (Int, error) {
	if i == None {
		return d, nil
	} else if res, err := Index(i); err != nil {
		return 0, err
	} else {
		return res, nil
	}
}

func computeNegativeIndex(index, length Int) Int {
	if index < 0 {
		index += length
	}
	return index
}

func computeBoundIndex(index, length Int) Int {
	if index < 0 {
		index = 0
	} else if index > length {
		index = length
	}
	return index
}

func computeRangeSlice(r *Range, s *Slice) (Object, error) {
	start, err := getIndexWithDefault(s.Start, 0)
	if err != nil {
		return nil, err
	}
	stop, err := getIndexWithDefault(s.Stop, r.Length)
	if err != nil {
		return nil, err
	}
	step, err := getIndexWithDefault(s.Step, 1)
	if err != nil {
		return nil, err
	}

	if step == 0 {
		return nil, ExceptionNewf(ValueError, "slice step cannot be zero")
	}
	start = computeNegativeIndex(start, r.Length)
	stop = computeNegativeIndex(stop, r.Length)

	start = computeBoundIndex(start, r.Length)
	stop = computeBoundIndex(stop, r.Length)

	startIndex := computeItem(r, start)
	stopIndex := computeItem(r, stop)
	stepIndex := step * r.Step

	var sliceLength Int
	if start < stop {
		if stepIndex < 0 {
			startIndex, stopIndex = stopIndex-1, startIndex-1
		}
	} else {
		if stepIndex < 0 {
			startIndex, stopIndex = stopIndex+1, startIndex+1
		}
	}
	sliceLength = computeRangeLength(startIndex, stopIndex, stepIndex)

	return &Range{
		Start:  startIndex,
		Stop:   stopIndex,
		Step:   stepIndex,
		Length: sliceLength,
	}, nil
}

// Check interface is satisfied
var _ I__getitem__ = (*Range)(nil)
var _ I__iter__ = (*Range)(nil)
var _ I_iterator = (*RangeIterator)(nil)

func (a *Range) M__eq__(other Object) (Object, error) {
	b, ok := other.(*Range)
	if !ok {
		return NotImplemented, nil
	}

	if a.Length != b.Length {
		return False, nil
	}

	if a.Length == 0 {
		return True, nil
	}
	if a.Start != b.Start {
		return False, nil
	}

	if a.Step == 1 {
		return True, nil
	}
	if a.Step != b.Step {
		return False, nil
	}

	return True, nil
}

func (a *Range) M__ne__(other Object) (Object, error) {
	b, ok := other.(*Range)
	if !ok {
		return NotImplemented, nil
	}

	if a.Length != b.Length {
		return True, nil
	}

	if a.Length == 0 {
		return False, nil
	}
	if a.Start != b.Start {
		return True, nil
	}

	if a.Step == 1 {
		return False, nil
	}
	if a.Step != b.Step {
		return True, nil
	}

	return False, nil
}

File: py/tests/range.py
# Copyright 2018 The go-python Authors.  All rights reserved.
# Use of this source code is governed by a BSD-style
# license that can be found in the LICENSE file.

doc="range"
a = range(255)
b = [e for e in a]
assert len(a) == len(b)
a = range(5, 100, 5)
b = [e for e in a]
assert len(a) == len(b)
a = range(100 ,0, 1)
b = [e for e in a]
assert len(a) == len(b)

a = range(100, 0, -1)
b = [e for e in a]
assert len(a) == 100
assert len(b) == 100

doc="range_get_item"
a = range(3)
assert a[2] == 2
assert a[1] == 1
assert a[0] == 0
assert a[-1] == 2
assert a[-2] == 1
assert a[-3] == 0

b = range(0, 10, 2)
assert b[4] == 8
assert b[3] == 6
assert b[2] == 4
assert b[1] == 2
assert b[0] == 0
assert b[-4] == 2
assert b[-3] == 4
assert b[-2] == 6
assert b[-1] == 8

doc="range_eq"
assert range(10) == range(0, 10)
assert not range(10) == 3
assert range(20) != range(10)
assert range(100, 200, 1) == range(100, 200)
assert range(0, 10, 3) == range(0, 12, 3)
assert range(2000, 100) == range(3, 1)
assert range(0, 10, -3) == range(0, 12, -3)
assert not range(0, 20, 2) == range(0, 20, 4)
try:
    range('3', 10) == range(2)
except TypeError:
    pass
else:
    assert False, "TypeError not raised"

doc="range_ne"
assert range(10, 0, -3) != range(12, 0, -3)
assert range(10) != 3
assert not range(100, 200, 1) != range(100, 200)
assert range(0, 10) != range(0, 12)
assert range(0, 10) != range(0, 10, 2)
assert range(0, 20, 2) != range(0, 21, 2)
assert range(0, 20, 2) != range(0, 20, 4)
assert not range(0, 20, 3) != range(0, 20, 3)
try:
    range('3', 10) != range(2)
except TypeError:
    pass
else:
    assert False, "TypeError not raised"

doc="range_str"
assert str(range(10)) == 'range(0, 10)'
assert str(range(10, 0, 3)) == 'range(10, 0, 3)'
assert str(range(0, 3)) == 'range(0, 3)'
assert str(range(10, 3, -2)) == 'range(10, 3, -2)'

doc="range_slice"
a = range(10)
assert a[::-1][0] == 9
assert a[::-1][9] == 0
assert a[0:3][0] == 0
assert a[0:3][2] == 2
assert a[-3:10][0] == 7
assert a[-100:13][0] == 0
assert a[-100:13][9] == 9

try:
    a[0:3][3]
except IndexError:
    pass
else:
    assert False, "IndexError not raised"
try:
    a[100:13][0]
except IndexError:
    pass
else:
    assert False, "IndexError not raised"
try:
    a[0:3:0]
except ValueError:
    pass
else:
    assert False, "ValueError not raised"

doc="range_index"
class Index:
    def __index__(self):
        return 1

a = range(10)
b = Index()
assert a[b] == 1
assert a[b:10] == a[1:10]
assert a[10:b:-1] == a[10:1:-1]

class NonIntegerIndex:
    def __index__(self):
        return 1.1

a = range(10)
b = NonIntegerIndex()
try:
    a[b]
except TypeError:
    pass
else:
    assert False, "TypeError not raised"

try:
    a[b:10]
except TypeError:
    pass
else:
    assert False, "TypeError not raised"

doc="finished"
